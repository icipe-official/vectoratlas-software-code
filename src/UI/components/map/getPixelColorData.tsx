export const getPixelColorData = (rgba:any) => {
  return ((rgba[0] + rgba[1] + rgba[2])*0.1);
};
