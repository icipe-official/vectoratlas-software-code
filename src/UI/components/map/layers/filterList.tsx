import React from 'react';
import { useDispatch } from 'react-redux';
import { useAppSelector } from '../../../state/hooks';
import ListItem from '@mui/material/ListItem';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import Collapse from '@mui/material/Collapse';
import ExpandLess from '@mui/icons-material/ExpandLess';
import ExpandMore from '@mui/icons-material/ExpandMore';
import LayersIcon from '@mui/icons-material/Layers';
import MapIcon from '@mui/icons-material/Map';
import FilterAltIcon from '@mui/icons-material/FilterAlt';
import FilterDropDown from './filterDropDown';
import MultipleFilterToggle from './filterToggle';
import { drawerListToggle, drawerToggle } from '../../../state/mapSlice';
import YearMonthPicker from './filterDatePicker';
import ThunderstormIcon from '@mui/icons-material/Thunderstorm';
import WbSunnyIcon from '@mui/icons-material/WbSunny';
import DataArrayIcon from '@mui/icons-material/DataArray';
import PestControlIcon from '@mui/icons-material/PestControl';
import EggIcon from '@mui/icons-material/Egg';
import DoneIcon from '@mui/icons-material/Done';
import CloseIcon from '@mui/icons-material/Close';
import EmojiNatureIcon from '@mui/icons-material/EmojiNature';
import BugReportIcon from '@mui/icons-material/BugReport';

export const FilterList = ({
  sectionTitle,
  sectionFlag,
}: {
  sectionTitle: string;
  overlays: Array<Object>;
  sectionFlag: string;
}) => {
  const dispatch = useDispatch();
  const open = useAppSelector((state) => state.map.map_drawer.open);
  const openNestList = useAppSelector((state) =>
    sectionFlag === 'overlays'
      ? state.map.map_drawer.overlays
      : sectionFlag === 'baseMap'
      ? state.map.map_drawer.baseMap
      : state.map.map_drawer.filters
  );
  const speciesList = useAppSelector((state) => state.map.filterValues.species);
  const countryList = useAppSelector(
    (state) => state.map.filterValues.countries
  );

  const ITEM_HEIGHT = 48;
  const ITEM_PADDING_TOP = 8;
  const MenuProps = {
    PaperProps: {
      style: {
        maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
        width: 250,
      },
    },
  };

  const handleClick = () => {
    if (open === true) {
      dispatch(drawerListToggle(sectionFlag));
    } else {
      dispatch(drawerToggle());
      dispatch(drawerListToggle(sectionFlag));
    }
  };
  return (
    <ListItem disablePadding sx={{ display: 'block' }}>
      <ListItemButton
        data-testid={`${sectionFlag}Button`}
        sx={{
          minHeight: 48,
          justifyContent: open ? 'initial' : 'center',
          px: 2.5,
        }}
        onClick={handleClick}
      >
        <ListItemIcon
          sx={{
            minWidth: 0,
            mr: open ? 3 : 'auto',
            justifyContent: 'center',
          }}
        >
          {sectionFlag === 'overlays' ? (
            <LayersIcon />
          ) : sectionFlag === 'baseMap' ? (
            <MapIcon />
          ) : (
            <FilterAltIcon />
          )}
        </ListItemIcon>
        <ListItemText primary={sectionTitle} sx={{ opacity: open ? 1 : 0 }} />
        {openNestList && open ? (
          <ExpandLess />
        ) : !openNestList && open ? (
          <ExpandMore />
        ) : (
          <></>
        )}
      </ListItemButton>
      <Collapse
        in={openNestList}
        timeout="auto"
        unmountOnExit
        sx={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          paddingX: 2,
          paddingBottom: 2,
          width: '100%',
        }}
      >
        <FilterDropDown filterTitle={'Country'} />
        <FilterDropDown filterTitle={'Species'} />
        <MultipleFilterToggle
          filterTitle={'Season'}
          filterToggleType={'string'}
          filterOptionsArray={[
            { name: 'wet', optionIcon: <ThunderstormIcon /> },
            { name: 'dry', optionIcon: <WbSunnyIcon /> },
            { name: 'empty', optionIcon: <DataArrayIcon /> },
          ]}
        />
        <MultipleFilterToggle
          filterTitle={'Control'}
          filterToggleType={'boolean'}
          filterOptionsArray={[
            { name: 'true', optionIcon: <DoneIcon /> },
            { name: 'false', optionIcon: <CloseIcon /> },
            { name: 'empty', optionIcon: <DataArrayIcon /> },
          ]}
        />
        <MultipleFilterToggle
          filterTitle={'Adult'}
          filterToggleType={'boolean'}
          filterOptionsArray={[
            { name: 'true', optionIcon: <EmojiNatureIcon /> },
            { name: 'false', optionIcon: <BugReportIcon /> },
            { name: 'empty', optionIcon: <DataArrayIcon /> },
          ]}
        />
        <MultipleFilterToggle
          filterTitle={'Larval'}
          filterToggleType={'boolean'}
          filterOptionsArray={[
            { name: 'true', optionIcon: <PestControlIcon /> },
            { name: 'false', optionIcon: <EggIcon /> },
            { name: 'empty', optionIcon: <DataArrayIcon /> },
          ]}
        />
        <YearMonthPicker filterTitle={'Time'} />
      </Collapse>
    </ListItem>
  );
};
export default FilterList;
