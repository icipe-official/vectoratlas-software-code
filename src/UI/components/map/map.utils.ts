export function unpackOverlays(map_layers: any) {
  if (map_layers.length === 0) {
    return [[], []];
  }
  const worldLayer = map_layers.find((l: any) => l.name === 'world');
  const overlayList = map_layers.filter((l: any) => l.name !== 'world');
  const worldMapLayers = worldLayer.overlays.map((o: any) => ({
    ...o,
    sourceLayer: worldLayer.name,
    sourceType: worldLayer.sourceType,
  }));
  return [overlayList, worldMapLayers];
}

export function responseToGEOJSON(occurrenceData: any) {
  const geoJSONPoints = (occurrenceData || []).map((d: any) => ({
    type: 'Feature',
    geometry: d.site.location,
    properties: {
      name: d.site.name,
      year_start: d.year_start,
      n_all: d.sample.n_all,
      species: d.recorded_species.species.species,
      series: d.recorded_species.species.series,
    },
  }));
  const geoJSONFeatureCollection = {
    type: 'FeatureCollection',
    features: geoJSONPoints,
  };
  return JSON.stringify(geoJSONFeatureCollection);
}
