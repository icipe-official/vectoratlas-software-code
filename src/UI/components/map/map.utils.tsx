export const getPixelColorData = (rgba: any) => {
  return (rgba[0] + rgba[1] + rgba[2]) * 0.1;
};

export function pixelHoverInteraction(
  e: any,
  layer: any,
  getDataFunction: Function,
  targetHTML: any
) {
  if (e.dragging) {
    return;
  }
  const pixelData = layer.getData(e.pixel);
  targetHTML.innerText = pixelData
    ? getDataFunction(pixelData).toFixed(2)
    : '0.00';
}

export function responseToGEOJSON(graphqlLocationObject){
  var pointGeoJSONArray = [];
  for (let i = 0; i < graphqlLocationObject.data.allGeoData.length; i++) {
    const graphqlPointObject = graphqlLocationObject.data.allGeoData[i];
    const pointGeoJSON = 
    { type: 'Feature',
      geometry: {
        type: 'Point', 
        coordinates: [graphqlPointObject.longitude, graphqlPointObject.latitude]
      },
      properties: {
        arbData: graphqlPointObject.value
      }
    };
    pointGeoJSONArray.push(pointGeoJSON);
  }
  const geoJSONFeatureCollection = 
  {type: 'FeatureCollection',
    features: pointGeoJSONArray
  };
  return JSON.stringify(geoJSONFeatureCollection);
}
