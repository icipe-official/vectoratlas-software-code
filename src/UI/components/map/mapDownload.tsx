import { Button, Link, Paper } from '@mui/material'
import React from 'react'

export default function MapDownload(): JSX.Element {

  return (
    <>

    <div id="map" className='map'></div>

   <Link 
        id="export-png" 
        className="btn btn-outline-dark" 
        role="button"
        underline='hover'
        variant='overline'
        fontSize='20px'
        style={
            {   display: 'flex', 
                justifyContent: 'center' ,
                
                
             }
        }>
            <i className="fa fa-download"></i> DOWNLOAD MAP IMAGE</Link>
    <Link id="image-download" download="map.png"></Link>

    </>
  )
};

