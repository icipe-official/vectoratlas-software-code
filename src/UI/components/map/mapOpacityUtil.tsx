// Opacity Control Functionality:  
export function updateOpacity(layer:any, opacityInput:any, opacityOutput:any) {
  const opacity = parseFloat(opacityInput.value);
  layer.setOpacity(opacity);
  opacityOutput.innerText = opacity.toFixed(2);
}
