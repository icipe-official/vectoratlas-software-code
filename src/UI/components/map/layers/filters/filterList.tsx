import React from 'react';
import { useAppSelector, useAppDispatch } from '../../../../state/hooks';
import ListItem from '@mui/material/ListItem';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import Collapse from '@mui/material/Collapse';
import ExpandLess from '@mui/icons-material/ExpandLess';
import ExpandMore from '@mui/icons-material/ExpandMore';
import ThunderstormIcon from '@mui/icons-material/Thunderstorm';
import WbSunnyIcon from '@mui/icons-material/WbSunny';
import DataArrayIcon from '@mui/icons-material/DataArray';
import FilterAltIcon from '@mui/icons-material/FilterAlt';
import DoneIcon from '@mui/icons-material/Done';
import CloseIcon from '@mui/icons-material/Close';
import EmojiNatureIcon from '@mui/icons-material/EmojiNature';
import BugReportIcon from '@mui/icons-material/BugReport';
import PestControlIcon from '@mui/icons-material/PestControl';
import EggIcon from '@mui/icons-material/Egg';
import FingerprintIcon from '@mui/icons-material/Fingerprint';
import HourglassEmptyIcon from '@mui/icons-material/HourglassEmpty';
import FilterDropDown from './filterDropDown';
import FilterToggle from './filterToggle';
import FilterToggleExclusive from './filterToggleExclusive';
import DateFilter from './dateFilter';
import { drawerListToggle, drawerToggle } from '../../../../state/map/mapSlice';
import Grid from '@mui/material/Grid';
import { AreaFilters } from './areaFilter';

export const FilterList = ({
  sectionTitle,
  sectionFlag,
}: {
  sectionTitle: string;
  sectionFlag: string;
}) => {
  const dispatch = useAppDispatch();
  const open = useAppSelector((state) => state.map.map_drawer.open);
  const openFilterPanel = useAppSelector(
    (state) => state.map.map_drawer.filters
  );

  const handleClick = () => {
    if (open === true) {
      dispatch(drawerListToggle(sectionFlag));
    } else {
      dispatch(drawerToggle());
      dispatch(drawerListToggle(sectionFlag));
    }
  };

  return (
    <ListItem disablePadding sx={{ display: 'block' }}>
      <ListItemButton
        data-testid={`${sectionFlag}Button`}
        sx={{
          minHeight: 48,
          justifyContent: open ? 'initial' : 'center',
          px: 2.5,
        }}
        onClick={handleClick}
      >
        <ListItemIcon
          sx={{
            minWidth: 0,
            mr: open ? 3 : 'auto',
            justifyContent: 'center',
          }}
        >
          <FilterAltIcon />
        </ListItemIcon>
        <ListItemText primary={sectionTitle} sx={{ opacity: open ? 1 : 0 }} />
        {openFilterPanel && open ? <ExpandLess /> : null}
        {!openFilterPanel && open ? <ExpandMore /> : null}
      </ListItemButton>
      <Collapse
        data-testid={`${sectionFlag}ListContainer`}
        in={openFilterPanel}
        timeout="auto"
        unmountOnExit
        sx={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          paddingX: 2,
          paddingBottom: 2,
          width: '100%',
        }}
      >
        <FilterDropDown filterTitle={'Country'} filterName="country" />
        <FilterDropDown
          filterTitle={'Species'}
          filterName="species"
          prefix="An. "
        />
        <Grid container spacing={3}>
          <Grid item md={6}>
            <FilterToggle
              filterTitle="Season"
              filterName="season"
              filterToggleType={'string'}
              filterOptionsArray={[
                { name: 'rainy', optionIcon: <ThunderstormIcon /> },
                { name: 'dry', optionIcon: <WbSunnyIcon /> },
                { name: 'empty', optionIcon: <DataArrayIcon /> },
              ]}
            />
          </Grid>
          <Grid item md={6}>
            <FilterToggle
              filterTitle="Control"
              filterName="control"
              filterToggleType={'boolean'}
              filterOptionsArray={[
                { name: 'true', optionIcon: <DoneIcon /> },
                { name: 'false', optionIcon: <CloseIcon /> },
                { name: 'empty', optionIcon: <DataArrayIcon /> },
              ]}
            />
          </Grid>
          <Grid item md={6}>
            <FilterToggle
              filterTitle="Adult"
              filterName="isAdult"
              filterToggleType={'boolean'}
              filterOptionsArray={[
                { name: 'true', optionIcon: <EmojiNatureIcon /> },
                { name: 'false', optionIcon: <BugReportIcon /> },
                { name: 'empty', optionIcon: <DataArrayIcon /> },
              ]}
            />
          </Grid>
          <Grid item md={6}>
            <FilterToggle
              filterTitle="Larval"
              filterName="isLarval"
              filterToggleType={'boolean'}
              filterOptionsArray={[
                { name: 'true', optionIcon: <PestControlIcon /> },
                { name: 'false', optionIcon: <EggIcon /> },
                { name: 'empty', optionIcon: <DataArrayIcon /> },
              ]}
            />
          </Grid>
          <Grid item md={12}>
            <FilterToggle
              filterTitle="Insecticide"
              filterName="insecticide"
              filterToggleType={'string'}
              filterOptionsArray={[
                { name: 'phenotypic', optionIcon: <FingerprintIcon /> },
                { name: 'genotypic', optionIcon: <HourglassEmptyIcon /> },
                { name: 'none', optionIcon: <DataArrayIcon /> },
              ]}
            />
          </Grid>
          <Grid item md={12}>
            <FilterToggleExclusive
              filterTitle="Include bionomics data:"
              filterName="includeBionomics"
            />
          </Grid>
        </Grid>
        <DateFilter filterTitle="Time" filterName="timeRange" />
        <AreaFilters />
      </Collapse>
    </ListItem>
  );
};
export default FilterList;
