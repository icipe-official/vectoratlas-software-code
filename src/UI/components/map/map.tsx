import React, { useRef, useEffect, useState } from 'react';

import Box from '@mui/material/Box';

import Map from 'ol/Map';
import View from 'ol/View';
import TileLayer from 'ol/layer/Tile';
import VectorLayer from 'ol/layer/Vector';
import VectorSource from 'ol/source/Vector';
import VectorTileLayer from 'ol/layer/VectorTile';
import VectorTileSource from 'ol/source/VectorTile';
import MVT from 'ol/format/MVT';
import { transform } from 'ol/proj';
import { Circle, Style, Fill, Stroke } from 'ol/style';
import XYZ from 'ol/source/XYZ';
import GeoJSON from 'ol/format/GeoJSON';
import Text from 'ol/style/Text';
import 'ol/ol.css';

import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { responseToGEOJSON } from './utils/map.utils';
import { getOccurrenceData, getSpeciesList } from '../../state/mapSlice';
import DrawerMap from './layers/drawerMap';

const defaultStyle = new Style({
  fill: new Fill({
    color: [0, 0, 0, 0],
  }),
  stroke: new Stroke({
    color: 'white',
    width: 0.5,
  }),
});

export const MapWrapper = () => {
  const mapStyles = useAppSelector((state) => state.map.map_styles);
  const occurrenceData = useAppSelector((state) => state.map.occurrence_data);
  const layerVisibility = useAppSelector((state) => state.map.map_overlays);
  const mapOverlays = useAppSelector((state) => state.map.map_overlays);

  const overlaysList = mapOverlays.filter(
    (l: any) => l.sourceLayer !== 'world'
  );

  const seriesArray = useAppSelector((state) => state.map.species_list);

  const dispatch = useAppDispatch();

  const [map, setMap] = useState<Map | null>(null);

  useEffect(() => {
    console.log('getting data');
    dispatch(getOccurrenceData());
    dispatch(getSpeciesList());
  }, [dispatch]);

  const layerStyles = Object.assign(
    {},
    ...mapStyles.layers.map((layer: any) => ({
      [layer.name]: new Style({
        fill: new Fill({
          color: layerVisibility.find((l: any) => l.name === layer.name)
            ?.isVisible
            ? layer.fillColor
            : [0, 0, 0, 0],
        }),
        stroke: layer.strokeColor
          ? new Stroke({
              color: layerVisibility.find((l: any) => l.name === layer.name)
                ?.isVisible
                ? layer.strokeColor
                : [0, 0, 0, 0],
              width: layer.strokeWidth,
            })
          : undefined,
        zIndex: layer.zIndex,
      }),
    }))
  );

  const mapElement = useRef(null);

  useEffect(() => {
    function markStyle(n_all: number, seriesString: string) {
      return new Style({
        image: new Circle({
          radius: 15,
          fill: new Fill({
            color: seriesArray.find((s: any) => s.series === seriesString)
              ?.color ?? [0, 0, 0, 0.7],
          }),
          stroke: new Stroke({
            color: '0',
            width: 1,
          }),
        }),
        text: new Text({
          text: n_all !== null ? String(n_all) : '',
          scale: 1.4,
          fill: new Fill({
            color: '#fff',
          }),
          stroke: new Stroke({
            color: '0',
            width: 1,
          }),
        }),
      });
    }

    function buildRasterLayer(layer: any) {
      const layerXYZ = new XYZ({
        url: `/data/${layer.name}/{z}/{x}/{y}.png`,
        maxZoom: 5,
      });
      const tileLayer = new TileLayer({
        preload: Infinity,
        source: layerXYZ,
        opacity: 1.0,
        visible: layerVisibility.find((l: any) => l.name === layer.name)
          ?.isVisible,
      });
      tileLayer.set('name', layer.name);

      return tileLayer;
    }

    const pointLayer = new VectorLayer({
      source: new VectorSource({
        features: new GeoJSON().readFeatures(
          responseToGEOJSON(occurrenceData),
          {
            featureProjection: 'EPSG:3857',
          }
        ),
      }),
      style: (feature) => {
        return markStyle(feature.get('n_all'), feature.get('series'));
      },
    });
    pointLayer.set('occurrence-data', true);

    const baseMapLayer = new VectorTileLayer({
      preload: Infinity,
      source: new VectorTileSource({
        attributions: 'Made with Natural Earth. cc Vector Atlas',
        format: new MVT(),
        maxZoom: 5,
        url: '/data/world/{z}/{x}/{y}.pbf',
      }),
      style: (feature) => {
        const layerName = feature.get('layer');
        return layerStyles[layerName] ?? defaultStyle;
      },
    });
    baseMapLayer.set('base-map', true);

    // Passing in layers to generate map with overlays
    const initialMap = new Map({
      target: 'mapDiv',
      layers: [
        baseMapLayer,
        ...overlaysList.map((l: any) => buildRasterLayer(l)),
        pointLayer,
      ],
      view: new View({
        center: transform([20, -5], 'EPSG:4326', 'EPSG:3857'),
        zoom: 4,
      }),
    });


    setMap(initialMap);

    // Initialise map
    return () => initialMap.setTarget(undefined);
  }, [mapStyles, seriesArray]);

  useEffect(() => {
    const pointsLayer = map
      ?.getAllLayers()
      .find((l) => l.get('occurrence-data'));

    pointsLayer?.setSource(
      new VectorSource({
        features: new GeoJSON().readFeatures(
          responseToGEOJSON(occurrenceData),
          {
            featureProjection: 'EPSG:3857',
          }
        ),
      })
    );
  }, [map, occurrenceData]);

  useEffect(() => {
    const allLayers = map?.getAllLayers();
    allLayers?.forEach((l) => {
      const matchingLayer = layerVisibility.find(
        (v) => v.name === l.get('name')
      );
      if (matchingLayer) {
        l.setVisible(matchingLayer.isVisible);
      }
    });

    const baseMapLayer = map
      ?.getAllLayers()
      .find((l) => l.get('base-map')) as VectorTileLayer;
    baseMapLayer?.setStyle((feature) => {
      const layerName = feature.get('layer');
      return layerStyles[layerName] ?? defaultStyle;
    });
  }, [map, layerVisibility, mapStyles]);

  useEffect(() => {
    document.getElementById('export-png')?.addEventListener('click', function () {
      map?.once('rendercomplete', function () {
        const mapCanvas = document.createElement('canvas');
        const size = map.getSize();
        if (!size || size.length < 2) {
          return;
        }
        mapCanvas.width = size[0];
        mapCanvas.height = size[1];
        const mapContext = mapCanvas.getContext('2d');
        if (!mapContext) {
          return;
        }
        Array.prototype.forEach.call(
          map.getViewport().querySelectorAll('.ol-layer canvas, canvas.ol-layer'),
          function (canvas) {
            if (canvas.width > 0) {
              const opacity =
                canvas.parentNode.style.opacity || canvas.style.opacity;
              mapContext.globalAlpha = opacity === '' ? 1 : Number(opacity);
              let matrix;
              const transform = canvas.style.transform;

              if (transform) {
                matrix = transform
                  .match(/^matrix\(([^\(]*)\)$/)[1]
                  .split(',')
                  .map(Number);
              } else {
                matrix = [
                  parseFloat(canvas.style.width) / canvas.width,
                  0,
                  0,
                  parseFloat(canvas.style.height) / canvas.height,
                  0,
                  0,
                ];
              }
              CanvasRenderingContext2D.prototype.setTransform.apply(
                mapContext,
                matrix
              );
              const backgroundColor = canvas.parentNode.style.backgroundColor;
              if (backgroundColor) {
                mapContext.fillStyle = backgroundColor;
                mapContext.fillRect(0, 0, canvas.width, canvas.height);
              }

              mapContext.drawImage(canvas, 0, 0);
            }
          }
        );
        mapContext.globalAlpha = 1;
        mapContext.setTransform(1, 0, 0, 1, 0, 0);
        const link = document.getElementById(
          'image-download'
        ) as HTMLAnchorElement | null;

        if (link != null) {
          link.href = mapCanvas.toDataURL();
          link.click();
        }
      });

        if(map?.getRenderer()){
          map?.renderSync();
        }


    });
  }, [map])


  return (
    <Box sx={{ display: 'flex', flexGrow: 1 }}>
      <DrawerMap />
      <Box component="main" sx={{ flexGrow: 1 }}>
        <div
          id="mapDiv"
          ref={mapElement}
          style={{ height: 'calc(100vh - 230px)' }}
          data-testid="mapDiv"
        ></div>
      </Box>
    </Box>
  );
};
