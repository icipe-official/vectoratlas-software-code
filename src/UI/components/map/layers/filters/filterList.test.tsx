import React from 'react';
import { fireEvent, render } from '../../../../test_config/render';
import { drawerListToggle, drawerToggle, initialState } from "../../../../state/map/mapSlice";
import { FilterList } from './filterList';

jest.mock('@mui/material/ListItemIcon', () => () => (<div>ListItemIcon icon</div>))
jest.mock('@mui/icons-material/ExpandLess', () => () => (<div>ExpandLess icon</div>))
jest.mock('@mui/material/ListItemText', () => (props) => (<div>List item text mock {JSON.stringify(props)}</div>))
jest.mock('@mui/material/ListItemButton', () => (props) => (
  <div onClick={props.onClick}>
    <div>List item button mock</div>
    {props.children}
  </div>
))

jest.mock('./filterDropDown', () => (props) => (<div>FilterDropDown mock {JSON.stringify(props)}</div>))
jest.mock('./filterToggle', () => (props) => (
  <div>FilterToggle mock {"title: " + props.filterTitle}{"name: " + props.filterName}{"toggle type: " + props.filterToggleType}</div>
))
jest.mock('./dateFilter', () => (props) => (<div>DateFilter mock {JSON.stringify(props)}</div>))

describe("FilterList", () => {
  let state;

  beforeEach(() => {
    state = { map: initialState() };
  })

  it('renders correctly when collapsed', () => {
    const { wrapper } = render(
      <FilterList sectionTitle="Filters" sectionFlag="filters" />,
      state
    );

    expect(wrapper.container).toMatchSnapshot();
  })

  it('renders correctly when expanded', () => {
    state.map.map_drawer.open = true;
    state.map.map_drawer.filters = true;

    const { wrapper } = render(
      <FilterList sectionTitle="Filters" sectionFlag="filters" />,
      state
    );

    expect(wrapper.container).toMatchSnapshot();
  })

  it('closes just the section if the drawer is open', async () => {
    state.map.map_drawer.open = true;
    state.map.map_drawer.filters = true;

    const { wrapper, store } = render(
      <FilterList sectionTitle="Filters" sectionFlag="filters" />,
      state
    );

    const sectionButton = await wrapper.findByText("List item button mock");
    fireEvent.click(sectionButton)

    const actions = store.getActions();
    expect(actions).toHaveLength(1);
    expect(actions[0]).toEqual(drawerListToggle('filters'))
  })

  it('opens the drawer and the section if the drawer is closed', async () => {
    state.map.map_drawer.open = false;

    const { wrapper, store } = render(
      <FilterList sectionTitle="Filters" sectionFlag="filters" />,
      state
    );

    const sectionButton = await wrapper.findByText("List item button mock");
    fireEvent.click(sectionButton)

    const actions = store.getActions();
    expect(actions).toHaveLength(2);
    expect(actions[0]).toEqual(drawerToggle())
    expect(actions[1]).toEqual(drawerListToggle('filters'))
  })
})