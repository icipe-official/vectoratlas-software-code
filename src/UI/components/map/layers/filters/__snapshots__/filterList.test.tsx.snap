// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterList renders correctly when collapsed 1`] = `
<div>
  <li
    class="MuiListItem-root MuiListItem-gutters css-uxa1ks-MuiListItem-root"
  >
    <div>
      <div>
        List item button mock
      </div>
      <div>
        ListItemIcon icon
      </div>
      <div>
        List item text mock 
        {"primary":"Filters","sx":{"opacity":0}}
      </div>
    </div>
  </li>
</div>
`;

exports[`FilterList renders correctly when expanded 1`] = `
<div>
  <li
    class="MuiListItem-root MuiListItem-gutters css-uxa1ks-MuiListItem-root"
  >
    <div>
      <div>
        List item button mock
      </div>
      <div>
        ListItemIcon icon
      </div>
      <div>
        List item text mock 
        {"primary":"Filters","sx":{"opacity":1}}
      </div>
      <div>
        ExpandLess icon
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pcvi2z-MuiCollapse-root"
      data-testid="filtersListContainer"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div>
            FilterDropDown mock 
            {"filterTitle":"Country","filterName":"country"}
          </div>
          <div>
            FilterDropDown mock 
            {"filterTitle":"Species","filterName":"species"}
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
            >
              <div>
                FilterToggle mock 
                title: Season
                name: season
                toggle type: string
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
            >
              <div>
                FilterToggle mock 
                title: Control
                name: control
                toggle type: boolean
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
            >
              <div>
                FilterToggle mock 
                title: Adult
                name: isAdult
                toggle type: boolean
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
            >
              <div>
                FilterToggle mock 
                title: Larval
                name: isLarval
                toggle type: boolean
              </div>
            </div>
          </div>
          <div>
            DateFilter mock 
            {"filterTitle":"Time","filterName":"timeRange"}
          </div>
        </div>
      </div>
    </div>
  </li>
</div>
`;
